Create tombstone module

Fabrizzio53 opened the pull request:
## Description

This new module allows users to **list**, **restore** objects from "Deleted Objects" container, and **delete** active objects.

- **Listing deleted objects:**  
  The module queries deleted objects using the LDAP filter `(isDeleted=TRUE)` along with the control `LDAP_SERVER_SHOW_DELETED_OID`, which is required to retrieve tombstoned entries. The query targets the DN `CN=Deleted Objects,DC=...`. By default, the first entry is skipped as it represents the container itself, not a real AD object (e.g., user or computer).

- **Restoring objects:**  
  To restore a deleted object, the module uses the ID obtained during the query and:
  - Removes the `isDeleted` attribute.
  - Replaces the `distinguishedName` with the value of `lastKnownParent`.

  This only works if the AD Recycle Bin is enabled and the user has the necessary permissions.

- **Deleting objects:**  
  The module also supports deleting objects to return to their "tombstoned form" using the `.delete()` method from `ldap3`. This is useful in scenarios where multiple objects with similar names exist (e.g., multiple versions of a user), and selective restoration or cleanup is needed.

## Type of change

- [x] New feature (non-breaking change which adds functionality)
 
## How Has This Been Tested?

Tested on a local lab that I created and HTB, tried to test at my work but I don't have the privs to do that ;)

## Screenshots (if appropriate):
![image](https://github.com/user-attachments/assets/b15cdacd-000f-4e4d-94b2-d59609b5c4a0)

After removing the user from this OU the query option (default) can be used to recover all the necessary informations from the Deleted Objects container.

![image](https://github.com/user-attachments/assets/7804e554-1c27-4bb4-8995-9139699601e6)

Right now the user has no way to confirm if he has or not privileges (this is something that in a future update a DACL parser can be added but I don't know how to do that now).

To restore an object from the Deleted Objects the user needs to call the "restore" action and pass the ID value recovered from the query action.

![image](https://github.com/user-attachments/assets/bd18fd79-5c28-43cb-9edd-09f9ade46c3d)

The first call to ldap fail since the user is not active, after the restore action the ldap connection worked again.

To delete the object the DN must be passed to the deleted action.

![image](https://github.com/user-attachments/assets/4eb0b884-495c-4ea3-9d2b-6522d1e53c4b)

## Checklist:

- [x] I have ran Ruff against my changes (via poetry: `poetry run python -m ruff check . --preview`, use `--fix` to automatically fix what it can)
- [x] If reliant on changes of third party dependencies, such as Impacket, dploot, lsassy, etc, I have linked the relevant PRs in those projects
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

## References

To perform this attack some actions must be performed at the domain:

https://nettools.net/how-to-delegate-object-restoration-rights/
https://learn.microsoft.com/pt-br/troubleshoot/windows-server/active-directory/non-administrators-view-deleted-object-container

By default only DA can restore and list objects from the Deleted Objects, I created this module after doing some machines from htb and vulnlab that AD Recycle Bin was active and you need to do a winrm / rdp session to execute the Restore-ADObject cmdlet, with this module you don't need to have a valid windows session anymore.

Fabrizzio53 committed:
Create gravedigger.py

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>

Fabrizzio53 commented on nxc/modules/gravedigger.py:
```diff
@@ -0,0 +1,258 @@
+            context.log.highlight(f"The object was not found with id {self.id}.")
+            return None
+
+        ldap_server, ldap_session = init_ldap_session(self.__domain, self.__username, self.__password, self.__lmhash, self.__nthash, self.__doKerberos, self.__host, self.__kdcHost, self.__aesKey, self.ssl)
```
Hi, I tried using the connection that netexec gives to me but as far as I know with impacket ldap I can't modify or delete ldap attributes so that's why I created a new ldap connection using ldap3

RajChowdhury240 commented:
tombstone module name would be ideal i guess

Fabrizzio53 commented:
I can change the name with no problem if everyone agrees

NeffIsBack commented:
Probably a less terrifying module name😅

Marshall-Hallenbeck commented:
I like the name :P

Fabrizzio53 committed:
Renamed gravedigger.py to tombstone.py and changed the name at the script

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>

Fabrizzio53 commented:
Name changed

RajChowdhury240 commented:
First of all amazing work mate! one thing i noticed

<img width="3162" height="1812" alt="image" src="https://github.com/user-attachments/assets/fc14e381-4f6c-4978-bd70-b1c9cab1fb05" />

it doesnt work with where machine only allows kerberos auth (NO NTLM) i used same syntax with `-k` option & it breaks , can you add kerberos support too ?

Fabrizzio53 committed:
update tombstone.py to fix when ntlm is disabled

When netexec impacket gets merged with the most up to date version from fortra this fix will work when a domain disabled ntlm

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>

Fabrizzio53 commented:
Done, but remember at the time that I'm writting this, impacket from netexec don't have the updated version, I know what machine you are doing and I used as a test example, if you want to test when the merge happens you need to modify this file:

/impacket/examples (line 241)

<img width="1089" height="304" alt="image" src="https://github.com/user-attachments/assets/b2b4cf9e-1c1e-49bc-bf1f-59b73b50002c" />

This is the same pr that I did for impacket to fix this problem when ntlm is disabled

Marshall-Hallenbeck commented:
New modules need to be added to the e2e tests

Fabrizzio53 committed:
Update e2e_commands.txt

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>

Marshall-Hallenbeck commented:
> Just to be sure I need to edit e2e_commands.txt right? But I think my fork has a different e2e because it is saying "This branch has conflicts that must be resolved"

Then resolve them?

Fabrizzio53 commented:
First time doing that, learned something new

NeffIsBack commented:
> but netexec uses another impacket, https://github.com/Pennyw0rth/impacket, when the pr that I made from impacket gets merged to the impacket that netexec use, then I can make the changes and the script will work again even when ntlm is not suported.

Gonna update our impacket fork, give me a bit

NeffIsBack commented:
Done, our impacket should be up-to-date now

NeffIsBack committed:
Merge branch 'main' into main

Fabrizzio53 committed:
Merge branch 'Pennyw0rth:main' into main

Fabrizzio53 committed:
Removed LDAP3 connection, removed old imports, removed SSL action

Removed the LDAP3 connection and now use impacket with the new CRUD method from the latest merged PR, removed the SSL option since now the first connection is handled by netexec and not ldap3

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>

Fabrizzio53 commented:
I decided to remove the ldap3 dependency and use the LDAP CRUD implementation built on impacket. Decided to rewrite some parts of the code too and removed the SSL options since now netexec handles a valid connection. From my domain the code is working, but if anyone finds a bug please let me know

Fabrizzio53 committed:
fixed ruff checks

fixed ruff checks

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+                resp = e.getAnswers()
+            else:
+                nxc_logger.debug(e)
+                return False
```
Please use the internal `connection.search` method which eliminates the needs for error handling, default attributes etc.

Fabrizzio53 replied:
removed all the LDAPSearchError checks and now it is using the connection.search

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+
+        for response in resp_parsed:
+            # The value 17 is the first entry from the ldap query when returning deleted objects and it should by default return the Deleted Objects OU information, by skipping this we return only objects that we want
+            if len(response) != 17 and self.id == response["name"].split(":")[1]:
```
I don't really get what this tries to exclude, could you provide an example screenshot? Why can't we skip this with an adjusted search filter?

Fabrizzio53 replied:
For the other issues I think I fixed, for this one to be honest I don't remember 100%, will do more testing at my lab and see If I can improve this part of the code

Fabrizzio53 replied:
I have updated the logic to handle the search results more robustly. Previously, the 'Deleted Objects' container was consistently the first entry returned and his value was always 17, but relying on that is not the safest option, because Microsoft might change the size and the script will just break.

```py
 for entries in resp_parsed:

   # This check ensures that we skip the result for the Default container and only get the result from the given ID.
        if "container" in entries["objectClass"] and entries["description"] == "Default container for deleted objects":
            continue
```

Instead of skipping the first index or checking the size, the code now explicitly checks the objectClass and description. This ensures we accurately identify and skip the default deleted objects container regardless of its position in the response or the size, allowing us to focus only on relevant objects like users, groups, and computers.

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+                context.log.highlight(f"ID      {response['name'].split(':')[1]}")
+                context.log.highlight(f"isDeleted       {response['isDeleted']}")
+                context.log.highlight(f"lastKnownParent       {response['lastKnownParent']}")
+                context.log.highlight("")
```
I think proper string formatting would be really helpful here for indentation, for an example see: https://github.com/Pennyw0rth/NetExec/blob/4998541b7576d2a6992de9c37bd52b1d177c55c7/nxc/protocols/ldap.py#L720-L727

Fabrizzio53 replied:
Improved on indentation by using a proper string formatting

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+                               "distinguishedName": [(MODIFY_REPLACE, [f"CN={self.__sAMAccountName},{self.__lastKnownParent}"])]  # Restore the user DN
+                               },
+                            controls=[show_deleted_control]
+            )
```
Indentation is a bit messed up here

Fabrizzio53 replied:
Fixed indentation

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+
+        if self.__sAMAccountName == "":
+            context.log.highlight(f"The object was not found with id {self.id}.")
+            return None
```
We should return only one variable type, e.g. boolean as done above. However, it looks like the return value is never used so just `return` would probably be sufficient.

Fabrizzio53 replied:
Removed "None" and now it uses False when the script does not find the object

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+        except LDAPSessionError as e:
+            context.log.highlight(f"Error at trying to recover the object {e}")
+
+        return None
```
See above

Fabrizzio53 replied:
Removed "None" and now it returns False when the object is not deleted

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+            context.log.highlight(f'Success "CN={self.__sAMAccountName},{self.__lastKnownParent}" restored')
+
+        except LDAPSessionError as e:
+            context.log.highlight(f"Error at trying to recover the object {e}")
```
That should probably be `.fail()`

Fabrizzio53 replied:
Indeed, changed from highlight to .fail()

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+
+        except LDAPSessionError as e:
+            context.log.highlight("")
+            context.log.highlight(f'Error when trying to delete "{self.deleteDN}" {e}')
```
Errors should be displayed with `.fail()`

Fabrizzio53 replied:
Indeed, changed from highlight to .fail()

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+            context.log.highlight("")
+            context.log.highlight(f'Error when trying to delete "{self.deleteDN}" {e}')
+
+        return
```
We don't need explicit returns at the end of a function in python

Fabrizzio53 replied:
Removed this return

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+                resp = e.getAnswers()
+            else:
+                nxc_logger.debug(e)
+                return False
```
Please use the `connection.search` function, see above

Fabrizzio53 replied:
removed all the LDAPSearchError checks and now it is using the connection.search

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+        if len(entries) < 2:
+            context.log.highlight("Recycle bin is not active on the domain or no user is in a tombstone state")
+
+            return None
```
see above

Fabrizzio53 replied:
Removed "None" and now it returns False if Recyle Bin container is not found on the domain

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+                context.log.highlight(f"ID      {response['name'].split(':')[1]}")
+                context.log.highlight(f"isDeleted       {response['isDeleted']}")
+                context.log.highlight(f"lastKnownParent       {response['lastKnownParent']}")
+                context.log.highlight("")
```
See above

Fabrizzio53 replied:
Improved on indentation by using a proper string formatting

NeffIsBack commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,218 @@
+        if self.action == "delete":
+            self.delete_object(context, connection)
+        if self.action == "restore":
+            self.restore_deleted_object(context, connection)
```
Imo this should be directly at the top of the module logic (because I now would need to search for the begin of the module), but I guess that is just taste.

NeffIsBack reviewed (changes requested):
Thanks for removing ldap3. 

A few things left to do, but overall the code looks already quite good :)

Fabrizzio53 committed:
fixed issues pointed on the PR

removed imports that were not being used, remove None returns, put the options at the start of the script and improved on indentation

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>

MTlyx commented:
There is a typo in the usage example for `nxc ldap -M tombstone --options` : the trailing `"` in `nxc ldap $DC-IP -u Username -p Password -M tombstone"` should be removed.

Fabrizzio53 committed:
Removed a typo " in the first example

Removed a typo " in the first example

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>

Fabrizzio53 commented:
Removed the typo

azoxlpf commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,168 @@
+            if "container" in entries["objectClass"] and entries["description"] == "Default container for deleted objects":
+                continue
+
+            context.log.highlight(f"{'sAMAccountName':<20}: {entries['sAMAccountName']}")
```
Not all tombstoned objects have `sAMAccountName` (OUs, contacts, etc.), so this raises `KeyError` and crashes the module, see the stack trace on query. Use `.get()` with a default (e.g. `entries.get('sAMAccountName', '')`) like that ([ldap.py](https://github.com/Pennyw0rth/NetExec/blob/main/nxc/protocols/ldap.py))  : 

```python
 self.logger.highlight(f"{'-Username-':<30}{'-Last PW Set-':<20}{'-BadPW-':<9}{'-Description-':<60}")
            for user in resp_parsed:
                pwd_last_set = user.get("pwdLastSet", "")
                if pwd_last_set:
                    pwd_last_set = "<never>" if pwd_last_set == "0" else datetime.fromtimestamp(self.getUnixTime(int(pwd_last_set))).strftime("%Y-%m-%d %H:%M:%S")
                # We default attributes to blank strings if they don't exist in the dict
                self.logger.highlight(f"{user.get('sAMAccountName', ''):<30}{pwd_last_set:<20}{user.get('badPwdCount', ''):<9}{user.get('description', ''):<60}")
                users.append(user.get("sAMAccountName", ""))
```

<img width="1893" height="1015" alt="Image" src="https://github.com/user-attachments/assets/b2948540-aa39-4a1b-a778-091bf0e22089" />

Fabrizzio53 replied:
code now uses context.log.highlight(f"{'sAMAccountName':<20}: {entries.get('sAMAccountName', '')}"), created an OU, removed and tested here, from the objects that I found, dn, ID, isDeleted and lastknowparent should be an attribute that is present on all types, but I could be wrong, in any case, this failsafe using .get() is being used at all attributes

azoxlpf commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,168 @@
+        self.action = "query"
+        self.id = ""
+        self.deleteDN = ""
+        if "ACTION" in module_options:
```
`ACTION` is not validated, a typo (e.g., ACTION=restor) will not match any branch in on_login, and the module will exit silently. Please validate it in `options()` using `.lower()` and a whitelist (query, restore, delete), and use `elif/else` with log.fail in `on_login` for unknown values. You can see [link_enable_cmdshell.py](https://github.com/Pennyw0rth/NetExec/blob/main/nxc/modules/link_enable_cmdshell.py) the pattern to follow

Fabrizzio53 replied:
Refactored the code to use .lower() and show an error message when an attribute that is not query, restore or delete gets passed

azoxlpf commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,168 @@
+                context.log.highlight(f"{'lastKnownParent':<20}: {entries['lastKnownParent']}")
+                context.log.highlight("")
+
+                self.__sAMAccountName = entries["sAMAccountName"]
```
Same `.get()` issue here, see above

Fabrizzio53 replied:
Applied the same fix from the other issue, the code now uses the .get() when grabbing the attributes

azoxlpf commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,168 @@
+    name = "tombstone"
+    description = "Query, restore and delete AD object"
+    supported_protocols = ["ldap"]
+    opsec_safe = True
```
This attribute is no used used by NXC, please remove it

Fabrizzio53 replied:
when I wrote this I remember seeing this being an option at the modules, in any case, removed opsec_safe and multiple_hosts

azoxlpf commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,168 @@
+    description = "Query, restore and delete AD object"
+    supported_protocols = ["ldap"]
+    opsec_safe = True
+    multiple_hosts = False
```
See above

Fabrizzio53 replied:
when I wrote this I remember seeing this being an option at the modules, in any case, removed opsec_safe and multiple_hosts

azoxlpf commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,168 @@
+    def restore_deleted_object(self, context, connection):
+
+        # ldap DN for deleted objects
+        dn = "CN=Deleted Objects," + self.domain_to_dn(self.__domain)
```
`domain_to_dn()` rebuilds the naming context from `connection.domain`, but LDAP already exposes the correct base via `connection.baseDN` (from rootDSE). Please use:

```python
dn = f"CN=Deleted Objects,{connection.baseDN}"
```

and drop `domain_to_dn()` + `self.__domain` if nothing else needs them.

Fabrizzio53 replied:
Wrote this because I didn't knew you had this method to get the baseDN, removed the domain_to_dn and now use the connection.baseDN, removed self.__domain too

azoxlpf commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,168 @@
+        self.__sAMAccountName = ""
+        self.__objectDN = ""
+        self.__lastKnownParent = ""
+        self.__domain = connection.domain
```
Duplicate assignment `self.__domain = connection.domain` is set twice, please drop one.

Fabrizzio53 replied:
removed both cases of self.__domain since it was not being used

azoxlpf commented on nxc/modules/tombstone.py:
```diff
@@ -0,0 +1,168 @@
+            context.log.highlight("No objects are in a tombstone state")
+            return False
+
+        context.log.highlight(f"Found {len(resp) - 1} deleted objects")
```
Hmm, I think `len(resp) - 1` doesn't match the number of objects actually displayed (the default container for deleted objects is filtered out in the loop). We should count the entries after filtering, or increment the value within the loop

Fabrizzio53 replied:
you are right, created a temporary var and increment for each value of deleted object that it founds, the number of objects is now shown at the end of the module

Fabrizzio53 committed:
Refactor tombstone

refactored the code to return empty sAMAccountName when object does not have this attribute.

removed some variables that were not being used, such as opsec, multiple_host and self.__domain.

removed internal domain to dn function to now use connection.baseDN.

the number of deleted objects should be right now, temporary var is created and incremented for each deleted object found.

Signed-off-by: Fabrizzio53 <88493503+Fabrizzio53@users.noreply.github.com>
